libbpf-tools/tcpconnlat.bpf.c

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "tcpconnlat.h"
+
+#define AF_INET    2
+#define AF_INET6   10
+
+const volatile __u64 targ_min_us = 0;
+const volatile pid_t targ_tgid = 0;
+
+struct piddata {
+	char comm[TASK_COMM_LEN];
+	u64 ts;
+	u32 tgid;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 4096);
+	__type(key, struct sock *);
+	__type(value, struct piddata);
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u32));
+} events SEC(".maps");
+
+static __always_inline int trace_connect(struct sock *sk)
+{
+	u32 tgid = bpf_get_current_pid_tgid() >> 32;
+	struct piddata piddata = {};
+
+	if (targ_tgid  && targ_tgid != tgid)
+		return 0;
+
+	bpf_get_current_comm(&piddata.comm, sizeof(piddata.comm));
+	piddata.ts = bpf_ktime_get_ns();
+	piddata.tgid = tgid;
+	bpf_map_update_elem(&start, &sk, &piddata, 0);
+	return 0;
+}
+
+SEC("fentry/tcp_v4_connect")
+int BPF_PROG(fentry__tcp_v4_connect, struct sock *sk)
+{
+	return trace_connect(sk);
+}
+
+SEC("kprobe/tcp_v6_connect")
```

Comment:
is this kprobe (and tcp_v4_connect is fentry) because it could be in a module?

Reply:
It's declared as `static`, so I have to use kprobe here. (: